File: frontend/src/api/entity.ts
import request from './request';

export interface EntityProperty {
    id: string;
    entity_id: string;
    prop_code: string;
    prop_name: string;
    prop_name_en: string | null;
    data_type: string;
    options_json: string[] | null;
    is_required: boolean;
    display_order: number;
    created_at: string;
    updated_at: string;
}

export interface Entity {
    id: string;
    entity_code: string;
    entity_name: string;
    entity_name_en: string | null;
    description: string | null;
    status: 'DRAFT' | 'ACTIVE';
    is_active: boolean;
    created_at: string;
    updated_at: string;
}

export interface EntityDetail extends Entity {
    properties: EntityProperty[];
}

export interface EntityListParams {
    page?: number;
    size?: number;
    status?: 'DRAFT' | 'ACTIVE';
    is_active?: boolean;
    search?: string;
}

export interface EntityListResponse {
    items: Entity[];
    total: number;
    page: number;
    size: number;
    pages: number;
}

export interface EntityCreateParams {
    entity_code: string;
    entity_name: string;
    entity_name_en?: string;
    description?: string;
}

export interface EntityUpdateParams {
    entity_name?: string;
    entity_name_en?: string;
    description?: string;
    status?: 'DRAFT' | 'ACTIVE';
}

export interface PropertyCreateParams {
    prop_code: string;
    prop_name: string;
    prop_name_en?: string;
    data_type?: string;
    options_json?: string[];
    is_required?: boolean;
    display_order?: number;
}

export interface PropertyUpdateParams {
    prop_name?: string;
    prop_name_en?: string;
    data_type?: string;
    options_json?: string[];
    is_required?: boolean;
    display_order?: number;
}

export const entityApi = {
    list: (params: EntityListParams) => {
        return request.get<any, EntityListResponse>('/entities', { params });
    },

    create: (data: EntityCreateParams) => {
        return request.post<any, Entity>('/entities', data);
    },

    get: (id: string) => {
        return request.get<any, EntityDetail>(`/entities/${id}`);
    },

    update: (id: string, data: EntityUpdateParams) => {
        return request.patch<any, Entity>(`/entities/${id}`, data);
    },

    delete: (id: string) => {
        return request.delete(`/entities/${id}`);
    },

    // Properties
    listProperties: (entityId: string) => {
        return request.get<any, EntityProperty[]>(`/entities/${entityId}/properties`);
    },

    createProperty: (entityId: string, data: PropertyCreateParams) => {
        return request.post<any, EntityProperty>(`/entities/${entityId}/properties`, data);
    },

    updateProperty: (entityId: string, propId: string, data: PropertyUpdateParams) => {
        return request.patch<any, EntityProperty>(`/entities/${entityId}/properties/${propId}`, data);
    },

    deleteProperty: (entityId: string, propId: string) => {
        return request.delete(`/entities/${entityId}/properties/${propId}`);
    },
};

File: app/schemas/entity.py
"""
Entity and EntityProperty Pydantic schemas.
"""
import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


# ============================================
# Entity Property Schemas
# ============================================

class EntityPropertyBase(BaseModel):
    """Base schema for entity property."""
    prop_code: str = Field(..., min_length=1, max_length=128)
    prop_name: str = Field(..., min_length=1, max_length=256)
    prop_name_en: str | None = Field(None, max_length=256)
    data_type: str = Field("STRING", pattern="^(STRING|INTEGER|FLOAT|BOOLEAN|ENUM)$")
    options_json: list[str] | None = None
    is_required: bool = False
    display_order: int = 0


class EntityPropertyCreate(EntityPropertyBase):
    """Schema for creating an entity property."""
    pass


class EntityPropertyUpdate(BaseModel):
    """Schema for updating an entity property."""
    prop_name: str | None = Field(None, max_length=256)
    prop_name_en: str | None = Field(None, max_length=256)
    data_type: str | None = Field(None, pattern="^(STRING|INTEGER|FLOAT|BOOLEAN|ENUM)$")
    options_json: list[str] | None = None
    is_required: bool | None = None
    display_order: int | None = None


class EntityPropertyResponse(EntityPropertyBase):
    """Schema for entity property response."""
    id: uuid.UUID
    entity_id: uuid.UUID
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# ============================================
# Entity Schemas
# ============================================

class EntityBase(BaseModel):
    """Base schema for entity."""
    entity_code: str = Field(..., min_length=1, max_length=128, pattern="^[a-zA-Z0-9_]+$")
    entity_name: str = Field(..., min_length=1, max_length=256)
    entity_name_en: str | None = Field(None, max_length=256)
    description: str | None = None


class EntityCreate(EntityBase):
    """Schema for creating an entity."""
    pass


class EntityUpdate(BaseModel):
    """Schema for updating an entity."""
    entity_name: str | None = Field(None, max_length=256)
    entity_name_en: str | None = Field(None, max_length=256)
    description: str | None = None
    status: str | None = Field(None, pattern="^(DRAFT|ACTIVE)$")


class EntityResponse(EntityBase):
    """Schema for entity response."""
    id: uuid.UUID
    status: str
    is_active: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class EntityDetailResponse(EntityResponse):
    """Schema for entity detail response with properties."""
    properties: list[EntityPropertyResponse] = []


class EntityListResponse(BaseModel):
    """Schema for paginated entity list."""
    items: list[EntityResponse]
    total: int
    page: int
    size: int
    pages: int
